### PR TITLE
Enable using File Watchers for cloud based Pipeline Roots

### PR DESCRIPTION
**Pull request by Sigmonia**
#### Rationale
https://docs.google.com/document/d/1adCI1KteCV-vNuCJ6Okz1lVcsO6AgZIKXZeSSg2cKQ0/edit?pli=1#

#### Related Pull Requests
* https://github.com/LabKey/cloud/pull/88
* https://github.com/LabKey/commonAssays/pull/373
* https://github.com/LabKey/MaxQuant/pull/40
* https://github.com/LabKey/premium/pull/276
* https://github.com/LabKey/targetedms/pull/435
* https://github.com/LabKey/nlp/pull/199
* https://github.com/LabKey/testAutomation/pull/883

#### Changes
* Lots of File --> nio.Path refactors
* Additional Cloud api interfaces
* new CloudWatcherThread to go along with the FileSystemWatcherImpl

### CODE REVIEW

File: api/src/org/labkey/api/pipeline/PipelineJob.java [hunk 1]
```diff
@@ -433,7 +433,6 @@ public PipeRoot getPipeRoot()
     public void setLogFilePath(Path logFile)
     {
         _logFilePathName = FileUtil.pathToString(logFile);
-//        _logJobStopStart.info("LogFilePath set to " + _logFilePathName);
```
Comment (labkey-klum) [inline api/src/org/labkey/api/pipeline/PipelineJob.java:436]: delete?

File: api/src/org/labkey/api/pipeline/PipelineProvider.java [hunk 2]
```diff
@@ -133,19 +151,20 @@ public boolean accept(File f, boolean checkSiblings)
...
-                                    {
-                                        return false;
-                                    }
-                                }
+                                        ft.isType(sibling.getFileName().toString()) &&
+                                        ft.getIndexMatch(sibling) < indexMatch);
+                            }
+                            catch (IOException e)
+                            {
                             }
```
Comment (labkey-klum) [inline api/src/org/labkey/api/pipeline/PipelineProvider.java:173]: should we be throwing an exception or at the least log an error?

File: api/src/org/labkey/api/files/FileSystemWatcherImpl.java [hunk 3]
```diff
@@ -314,6 +319,68 @@ private void close()
...
+                {
+                        CloudStoreService.get().executeWatchJob(cloudWatcherJob);
+                });
+            }
+            catch (Throwable e)  // Make sure throwables don't kill the background thread
+            {
+                ExceptionUtil.logExceptionToMothership(null, e);
+            }
+
+            Thread.sleep(10000);
```
Comment (labkey-klum) [inline api/src/org/labkey/api/files/FileSystemWatcherImpl.java]: Similar to the `FileSystemWatcherThread` would it be possible to move from looping/waiting every 10 seconds? I don't know whether we can get notified by the SQS service and placing some object in a blocking queue.

**Review by labkey-klum — APPROVED** [review]
see inline comments

File: api/src/org/labkey/api/pipeline/PipelineJob.java (hunk 1 above)
Comment (Sigmonia) [reply api/src/org/labkey/api/pipeline/PipelineJob.java:436]: deleted.

File: api/src/org/labkey/api/pipeline/PipelineProvider.java (hunk 2 above)
Comment (Sigmonia) [reply api/src/org/labkey/api/pipeline/PipelineProvider.java:173]: Added Logging and returning false, as I think it better matches the previous behavior.

File: api/src/org/labkey/api/files/FileSystemWatcherImpl.java (hunk 3 above)
Comment (Sigmonia) [reply api/src/org/labkey/api/files/FileSystemWatcherImpl.java]: This is the difference between SQS and SNS. SQS has no notification system, only polling I did switch it to an explicit timer task. It is now using the same period constant as the pollingQueue.

**Pull request merged**